server-spi/src/main/java/org/keycloak/models/ClientModel.java

```diff
@@ -34,10 +36,33 @@
     String PUBLIC_KEY = "publicKey";
     String X509CERTIFICATE = "X509Certificate";
 
+    interface ClientCreationEvent extends ProviderEvent {
+        ClientModel getCreatedClient();
+    }
+
+    // Called also during client creation after client is fully initialized (including all attributes etc)
+    interface ClientUpdatedEvent extends ProviderEvent {
+        ClientModel getUpdatedClient();
+        KeycloakSession getKeycloakSession();
+    }
+
+    interface ClientRemovedEvent extends ProviderEvent {
+        ClientModel getClient();
+        KeycloakSession getKeycloakSession();
+    }
+
     /**
-     * Stores the current state of the client immediately to the underlying store, similarly to a commit.
+     * Notifies other providers that this client has been updated.
+     * <p>
+     * After a client is updated, providers can register for {@link ClientUpdatedEvent}.
+     * The setters in this model do not send an update for individual updates of the model.
+     * This method is here to allow for sending this event for this client,
+     * allowsing for to group multiple changes of a client and signal that
+     * all the changes in this client have been performed.
      *
-     * @deprecated Do not use, to be removed
+     * @see ProviderEvent
+     * @see ProviderEventManager
+     * @see ClientUpdatedEvent
      */
     void updateClient();
```

Comment:
@hmlnarik Do you know about any scenario/place where this method is really necessary and it is not possible to make it work in the way that everything is committed and everyone notified at the end of the transaction? 

To me this still seems like something we should get rid of eventually, I think the code would be easier to understand and also debugging would be easier without this idea of notifying/flushing in the middle of transaction. I understand it is not possible now, as apparently there is a lot of code relying on it, but we should not support using it, therefore I would leave the `@deprecated` comment as is or at least I would add some warning to the javadoc that this should be used only in case it is really necessary.

Reply:
You are right, I missed the removal of this comment. Thanks.
When walking through the code, I believe this should be possible to remove by publishing the event before the transaction end. I leave this to https://issues.redhat.com/browse/KEYCLOAK-16021, since this should be actually a separate issue. Introduced https://issues.redhat.com/browse/KEYCLOAK-16077.